### ISSUE TITLE
libbpf-tools: tcpconnect ARM32 issues

### ISSUE THREAD

**Issue by jvijtiuk** [iovisor/bcc#3044]
Hello,

while testing tcpconnect on ARM32, we (@jpetrina, @jsmolic and me) encountered various issues. We have primarily tested on the ALFA Network AP120C-AC which is a device based on IPQ4018 SoC ARM. The kernel version there was 5.7.

The first issue that we encountered on the device was the use of __int128 for IPv6 addresses. __int128 isn't supported on 32 bit architectures [0][1]. The issue can also be reproduced on x86_64 systems with the following commands:

```
[user@pc ~]$ echo '__int128 a;' > test.c
[user@pc ~]$ gcc test.c
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../lib/Scrt1.o: in function '_start':
(.text+0x24): undefined reference to 'main'
collect2: error: ld returned 1 exit status
[user@pc ~]$ gcc -m32 test.c
test.c:1:1: error: ‘__int128’ is not supported on this target
    1 | __int128 a;
       | ^~~~~~~~
[user@pc ~]$
```

The same issue appears on ARM32. One solution would be to check for ```__SIZEOF_INT128__``` and if it is not available use an array of 4 one byte characters or something similar.

However, we did not manage to fix the issue, as we ran into other issues on ARM32 which appear even with all the IPv6 code removed.
On 32 bit platforms, the compiler seems to use 64 bit pointers in BPF code, and 32 bit values for the size of
pointer values in maps, which then causes the BPF verifier to report errors.
I haven't managed to look into how BPF works on 32 bit platforms yet, so the assumption might be wrong.
The line that was causing issues is [tcpconnect.bpf.c:171](https://github.com/iovisor/bcc/blob/9ef20e7aa61f6a4d7576f906fdaf32329bec34bc/libbpf-tools/tcpconnect.bpf.c#L171) 

The issue seems to be that skpp is interpeted as a u64 *, while the map value size is interpreted as a u32 *, thus the 4 byte map value size in the output below.

The line generates the following BPF assembly:  " 15: r7 = *(u64 *)(r0 + 0)", which then causes the following verifier error:
```
libbpf: -- BEGIN DUMP LOG ---
libbpf:
Unrecognized arg#0 type PTR
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
0: (bf) r6 = r1
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
1: (79) r7 = *(u64 *)(r6 +0)
; __u64 pid_tgid = bpf_get_current_pid_tgid();
2: (85) call bpf_get_current_pid_tgid#14
3: (bf) r8 = r0
; __u32 tid = pid_tgid;
4: (63) *(u32 *)(r10 -52) = r8
5: (bf) r2 = r10
;
6: (07) r2 += -52
; skpp = bpf_map_lookup_elem(&sockets, &tid);
7: (18) r1 = 0xcb7d8c00
9: (85) call bpf_map_lookup_elem#1
; if (!skpp)
10: (15) if r0 == 0x0 goto pc+130
 R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv(id=0) R10=fp0 fp-56=mmmm????
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
11: (67) r7 <<= 32
12: (77) r7 >>= 32
; if (ret)
13: (55) if r7 != 0x0 goto pc+122
 R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R6=ctx(id=0,off=0,imm=0) R7_w=inv0 R8=inv(id=0) R10=fp0 fp-56=mmmm????
14: (b7) r1 = 12
; sk = *skpp;
15: (79) r7 = *(u64 *)(r0 +0)
 R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R1_w=inv12 R6=ctx(id=0,off=0,imm=0) R7_w=inv0 R8=inv(id=0) R10=fp0 fp-56=mmmm????
invalid access to map value, value_size=4 off=0 size=8
R0 min value is outside of the array range
processed 15 insns (limit 1000000) max_states_per_insn 0 total_states 1 peak_states 1 mark_read 1

libbpf: -- END LOG --
libbpf: failed to load program 'kretprobe/tcp_v4_connect'
libbpf: failed to load object 'tcpconnect_bpf'
libbpf: failed to load BPF skeleton 'tcpconnect_bpf': -4007
failed to load BPF object: -4007
```

Changing the line to "sk = *((u32 *) skpp);" seems to force 32 bit access, and stops the verifier complaining about that issue, but the program still wouldn't load on our device.

Would the cast to u32 pointer make sense, or are we missing something?

The verifier output after we used the u32 pointer cast looked like this:
```
libbpf: load bpf program failed: ERROR: strerror_r(524)=22
libbpf: -- BEGIN DUMP LOG ---
libbpf:
Unrecognized arg#0 type PTR
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
0: (bf) r6 = r1
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
1: (79) r7 = *(u64 *)(r6 +0)
; __u64 pid_tgid = bpf_get_current_pid_tgid();
2: (85) call bpf_get_current_pid_tgid#14
3: (bf) r8 = r0
; __u32 tid = pid_tgid;
4: (63) *(u32 *)(r10 -52) = r8
5: (bf) r2 = r10
;
6: (07) r2 += -52
; skpp = bpf_map_lookup_elem(&sockets, &tid);
7: (18) r1 = 0xcc8b6000
9: (85) call bpf_map_lookup_elem#1
; if (!skpp)
10: (15) if r0 == 0x0 goto pc+130
 R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv(id=0) R10=fp0 fp-56=mmmm????
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
11: (67) r7 <<= 32
12: (77) r7 >>= 32
; if (ret)
13: (55) if r7 != 0x0 goto pc+122
 R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R6=ctx(id=0,off=0,imm=0) R7_w=inv0 R8=inv(id=0) R10=fp0 fp-56=mmmm????
14: (b7) r1 = 12
; sk = *((__u32 *)skpp);
15: (61) r7 = *(u32 *)(r0 +0)
 R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R1_w=inv12 R6=ctx(id=0,off=0,imm=0) R7_w=inv0 R8=inv(id=0) R10=fp0 fp-56=mmmm????
16: (bf) r3 = r7
17: (0f) r3 += r1
18: (bf) r1 = r10
;
19: (07) r1 += -54
; BPF_CORE_READ_INTO(&dport, sk, __sk_common.skc_dport);
20: (b7) r2 = 2
21: (85) call bpf_probe_read#4
last_idx 21 first_idx 10
regs=4 stack=0 before 20: (b7) r2 = 2
; if (filter_port(dport))
22: (69) r9 = *(u16 *)(r10 -54)
; if (filter_ports_len == 0)
23: (18) r1 = 0xd1141004
25: (61) r2 = *(u32 *)(r1 +0)
 R0=inv(id=0) R1_w=map_value(id=0,off=4,ks=4,vs=272,imm=0) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R8=inv(id=0) R9_w=inv(id=0,umax_value=65535,?
; if (filter_ports_len == 0)
26: (15) if r2 == 0x0 goto pc+20
last_idx 26 first_idx 22
regs=4 stack=0 before 25: (61) r2 = *(u32 *)(r1 +0)
; if (do_count) {
47: (18) r1 = 0xd1141000
49: (71) r1 = *(u8 *)(r1 +0)
 R0=inv(id=0) R1_w=map_value(id=0,off=0,ks=4,vs=272,imm=0) R2_w=invP0 R6=ctx(id=0,off=0,imm=0) R7=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R8=inv(id=0) R9_w=inv(id=0,umax_v?
; if (do_count) {
50: (15) if r1 == 0x0 goto pc+43
last_idx 50 first_idx 22
regs=2 stack=0 before 49: (71) r1 = *(u8 *)(r1 +0)
;
94: (77) r8 >>= 32
95: (b7) r1 = 0
; struct event event = {};
96: (7b) *(u64 *)(r10 -16) = r1
last_idx 96 first_idx 22
regs=2 stack=0 before 95: (b7) r1 = 0
97: (7b) *(u64 *)(r10 -8) = r1
98: (7b) *(u64 *)(r10 -24) = r1
99: (7b) *(u64 *)(r10 -32) = r1
100: (7b) *(u64 *)(r10 -40) = r1
101: (7b) *(u64 *)(r10 -48) = r1
; event.pid = pid;
102: (63) *(u32 *)(r10 -16) = r8
; event.uid = bpf_get_current_uid_gid();
103: (85) call bpf_get_current_uid_gid#15
; event.uid = bpf_get_current_uid_gid();
104: (63) *(u32 *)(r10 -12) = r0
; event.ts_us = bpf_ktime_get_ns() / 1000;
105: (85) call bpf_ktime_get_ns#5
; event.ts_us = bpf_ktime_get_ns() / 1000;
106: (37) r0 /= 1000
; event.ts_us = bpf_ktime_get_ns() / 1000;
107: (7b) *(u64 *)(r10 -24) = r0
108: (b7) r1 = 4
109: (bf) r3 = r7
110: (0f) r3 += r1
111: (bf) r8 = r10
;
112: (07) r8 += -48
; BPF_CORE_READ_INTO(&event.saddr_v4, sk, __sk_common.skc_rcv_saddr);
113: (bf) r1 = r8
114: (b7) r2 = 4
115: (85) call bpf_probe_read#4
last_idx 115 first_idx 104
regs=4 stack=0 before 114: (b7) r2 = 4
116: (b7) r1 = 0
117: (0f) r7 += r1
; BPF_CORE_READ_INTO(&event.daddr_v4, sk, __sk_common.skc_daddr);
118: (bf) r1 = r10
119: (07) r1 += -44
120: (b7) r2 = 4
121: (bf) r3 = r7
122: (85) call bpf_probe_read#4
last_idx 122 first_idx 116
regs=4 stack=0 before 121: (bf) r3 = r7
regs=4 stack=0 before 120: (b7) r2 = 4
; event.dport = dport;
123: (6b) *(u16 *)(r10 -8) = r9
; struct event event = {};
124: (bf) r1 = r10
125: (07) r1 += -40
; bpf_get_current_comm(event.task, sizeof(event.task));
126: (b7) r2 = 16
127: (85) call bpf_get_current_comm#16
last_idx 127 first_idx 116
regs=4 stack=0 before 126: (b7) r2 = 16
; bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
128: (bf) r1 = r6
129: (18) r2 = 0xce6ccc00
131: (18) r3 = 0xffffffff
133: (bf) r4 = r8
134: (b7) r5 = 48
135: (85) call bpf_perf_event_output#25
last_idx 135 first_idx 128
regs=20 stack=0 before 134: (b7) r5 = 48
136: (bf) r2 = r10
;
137: (07) r2 += -52
; bpf_map_delete_elem(&sockets, &tid);
138: (18) r1 = 0xcc8b6000
140: (85) call bpf_map_delete_elem#3
; int BPF_KRETPROBE(kretprobe__tcp_v4_connect, int ret)
141: (b7) r0 = 0
142: (95) exit

from 13 to 136: R0=map_value(id=0,off=0,ks=4,vs=4,imm=0) R6=ctx(id=0,off=0,imm=0) R7_w=inv(id=0,umax_value=2147483647,var_off=(0x0; 0x7fffffff)) R8=inv(id=0) R10=fp0 fp-56=mmmm????
; bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
136: (bf) r2 = r10
;
137: (07) r2 += -52
; bpf_map_delete_elem(&sockets, &tid);
138: (18) r1 = 0xcc8b6000
140: (85) call bpf_map_delete_elem#3
141: safe

from 10 to 141: safe
processed 80 insns (limit 1000000) max_states_per_insn 0 total_states 6 peak_states 6 mark_read 5

libbpf: -- END LOG --
libbpf: failed to load program 'kretprobe/tcp_v4_connect'
libbpf: failed to load object 'tcpconnect_bpf'
libbpf: failed to load BPF skeleton 'tcpconnect_bpf': -4007
failed to load BPF object: -4007
```

I'm not sure what exactly the issue there is, as I can't see an error. Following that, we reduced the tcpconnect example to this minimal example:
```
include "vmlinux.h"

#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>

SEC("kprobe/tcp_v4_connect")
int BPF_KPROBE(kprobe__tcp_v4_connect, struct sock *sk)
{
	bpf_printk("Tcp connection established\n");
        return 0;
}


char LICENSE[] SEC("license") = "GPL";

```

Running that causes a kernel panic on our AP120C device. The issue seems to be caused by a buffer overflow in memcpy that occurs while kprobes are being set up. We are still investigating and reproducing the issue on other ARM32 systems. We have also tested with the minimal example on a Raspberry Pi 2, and got a kernel oops, caused by the same underlying memcpy issue. The Raspberry Pi 2 was running a 5.4 kernel and has a Broadcom BCM2836 SoC with a quad-core Cortex-A7.

In the meantime I've tried to reproduce the issues on a QEMU ARM32 virt board running Debian with 4.19 kernel, as I haven't managed to get 5.7 working there yet. The minimal example doesn't cause an oops or panic on that system, but it still doesn't work, while other BPF samples do work. The output looks like this:

```
root@debian:/home/user/2/usr/bin# ./tcp_connect2
libbpf: load bpf program failed: Invalid argument
libbpf: failed to load program 'kprobe/tcp_v4_connect'
libbpf: failed to load object 'tcpconnect2_bpf'
libbpf: failed to load BPF skeleton 'tcpconnect2_bpf': -22
failed to load BPF object -22
```

In any case, we do not think that the kernel issues are caused primarily by tcpconnect, as it seems like a general kernel kprobe issue, so we will report that on the kernel mailing lists, but are sharing them here in case someone runs into the same issues while debugging other tcpconnect issues on ARM.

If there is any more information we can provide to help fix the issues and get tcpconnect working on ARM32 please let us know.

Thanks,
Juraj

[0] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60846
[1] https://stackoverflow.com/questions/16088282/is-there-a-128-bit-integer-in-gcc

**Comment by jvijtiuk** [comment]
@anakryiko @aspsk could you please take a look at this? Thanks.

**Comment by anakryiko** [comment]
> Hello,
> 
> while testing tcpconnect on ARM32, we (@jpetrina, @jsmolic and me) encountered various issues. We have primarily tested on the ALFA Network AP120C-AC which is a device based on IPQ4018 SoC ARM. The kernel version there was 5.7.
> 
> The first issue that we encountered on the device was the use of __int128 for IPv6 addresses. __int128 isn't supported on 32 bit architectures [0][1]. The issue can also be reproduced on x86_64 systems with the following commands:

I don't think why `__int128` was used in the first place, it's never really used as an integer. So I think just defining it as a char[16] would be absolutely fine.


> On 32 bit platforms, the compiler seems to use 64 bit pointers in BPF code, and 32 bit values for the size of
> pointer values in maps, which then causes the BPF verifier to report errors.

No, BPF pointers are always 64-bit values. The problem here is actually libbpf. If built for 32-bit architecture, it will resolve `struct sock *` in `__type(value, struct sock*);` to a 32-bit (because that's the size of pointer on host system). That's completely wrong, libbpf should have created map with value_size = 8. I'll need to think how to fix this properly, but I'll post a fix ASAP.

> libbpf: load bpf program failed: ERROR: strerror_r(524)=22

524 is -ENOTSUPP, it's not clear where from and why this is returned, but the code verification itself is successful, as can be seen from:

> from 10 to 141: safe
> processed 80 insns (limit 1000000) max_states_per_insn 0 total_states 6 peak_states 6 mark_read 5

> 
> Running that causes a kernel panic on our AP120C device. The issue seems to be caused by a buffer overflow in memcpy that occurs while kprobes are being set up. We are still investigating and reproducing the issue on other ARM32 systems. We have also tested with the minimal example on a Raspberry Pi 2, and got a kernel oops, caused by the same underlying memcpy issue. The Raspberry Pi 2 was running a 5.4 kernel and has a Broadcom BCM2836 SoC with a quad-core Cortex-A7.

That might be related to ftrace/kprobe sub-system and might be very ARM specific, as you suggested below.

> 
> In the meantime I've tried to reproduce the issues on a QEMU ARM32 virt board running Debian with 4.19 kernel, as I haven't managed to get 5.7 working there yet. The minimal example doesn't cause an oops or panic on that system, but it still doesn't work, while other BPF samples do work. The output looks like this:
> 
> ```
> root@debian:/home/user/2/usr/bin# ./tcp_connect2
> libbpf: load bpf program failed: Invalid argument
> libbpf: failed to load program 'kprobe/tcp_v4_connect'
> libbpf: failed to load object 'tcpconnect2_bpf'
> libbpf: failed to load BPF skeleton 'tcpconnect2_bpf': -22
> failed to load BPF object -22
> ```

-EINVAL is super generic, I'm not sure how I can help. If you know how to use function traces (I don't, but I've seen examples and they show which function calls which other function inside the kernel), that would allow to know where exactly in bpf() syscall we returned error, hopefully.

**Comment by anakryiko** [comment]
> -EINVAL is super generic, I'm not sure how I can help. If you know how to use function traces (I don't, but I've seen examples and they show which function calls which other function inside the kernel), that would allow to know where exactly in bpf() syscall we returned error, hopefully.

Ok, so I've just successfully tried that with help of perf. Try something like this:

```
sudo perf ftrace -G 'bpf_check' ./tcpconnect
```

This would give a very long list of nested function calls, but it should show what's going on. If you can show the result for the above case where you were getting -ENOTSUPP, that would be great! Just don't paste it inline here, use gist.github.com or something :)

**Comment by aspsk** [comment]
> > Hello,
> > while testing tcpconnect on ARM32, we (@jpetrina, @jsmolic and me) encountered various issues. We have primarily tested on the ALFA Network AP120C-AC which is a device based on IPQ4018 SoC ARM. The kernel version there was 5.7.
> > The first issue that we encountered on the device was the use of __int128 for IPv6 addresses. __int128 isn't supported on 32 bit architectures [0][1]. The issue can also be reproduced on x86_64 systems with the following commands:
> 
> I don't think why `__int128` was used in the first place, it's never really used as an integer. So I think just defining it as a char[16] would be absolutely fine.

@jvijtiuk @anakryiko See https://github.com/iovisor/bcc/pull/3051 

I copied the usage of `__int128` to be consistent with the original `tcpconnect.py`. In fact, all original tools use `__int128`, so they need to be fixed at some point.

**Comment by anakryiko** [comment]
@jvijtiuk, the issue with wrong map value size is going to be fixed by https://patchwork.ozlabs.org/project/netdev/patch/20200813071722.2213397-4-andriin@fb.com/ soon.

**Comment by jvijtiuk** [comment]
Firstly, thank you all for the really quick fixes and sorry for the long delay.

> @jvijtiuk, the issue with wrong map value size is going to be fixed by 
> https://patchwork.ozlabs.org/project/netdev/list/?series=195352&state=* soon.

Great, thanks. I see that the fixes have already been merged. We are currently testing that on the AP120C device. I will get back to you with the results tomorrow.

> Ok, so I've just successfully tried that with help of perf. Try something like this:
> 
> ```
> sudo perf ftrace -G 'bpf_check' ./tcpconnect
> ```
[Here](https://gist.github.com/jvijtiuk/932054499ffeb211793e478d211c34bd) is the gist with the results of the perf command. From what I understand the trace points to the fixup_bpf_calls function returning EINVAL, as the rest of the code after that call is cleanup code that is always called. I have enabled additional kernel verifier logs by calling bpf_object__load_xattr instead of the usual CO-RE skeleton load call, but I don't get any additional verifier output in QEMU, while on x86 I do see some additional logs.


Looking at the source of fixup_bpf_calls, there seem to be four places where it could return EINVAL, the first when something goes wrong with BPF_LD, second when bpf_map_ptr_poisoned fails, third when map_gen_lookup fails, and the fourth has something to do with map_poke. All the errors except the second one should show an error message that the bpf verifier is misconfigured. However, as I can't get any verifier logs to be printed, I'm not sure which error might be causing the EINVAL return. As there are no tail calls the second error doesn't make much sense. No maps are used in the minimized tcpconnect example so that should rule out errors three and four, which would mean that there is something wrong with the BPF_LD part.

Here is the result of uname -a for the QEMU system I'm using:
```
Linux debian 4.19.0-10-armmp-lpae #1 SMP Debian 4.19.132-1 (2020-07-24) armv7l GNU/Linux
```

[Here](https://gist.github.com/jvijtiuk/b3e25bd850d4be481f15af9211edbaed) is the kernel config on the system. As far as I can tell, nothing major seems to be missing. CONFIG_BPF_JIT_ALWAYS_ON was not set, so i tested with ```/proc/sys/net/core/bpf_jit_enable``` set to 0 and 1, with no differences in behaviour.
Additionally, I've managed to test the issue with ply by running the following command:
```
ply 'kprobe:tcp_v4_connect {print("hello");}'
```

That causes the same kernel oops in memcpy that we noticed on other arm32 systems, so I assume that the EINVAL issue is not due to the kernel config, since the ply command should have the same functionality of the minimal tcpconnect example I originally posted.

If there is any additional information that I can provide please let me know.

**Comment by jvijtiuk** [comment]
> We are currently testing that on the AP120C device. I will get back to you with the results tomorrow.
The patches to libbpf seem to have solved the previous verifier issue with 32/64 bit pointers.

Now we get the following results with ENOTSUPP again:
```
libbpf: Error in bpf_create_map_xattr(sockets):Invalid argument(-22). Retrying without BTF.
libbpf: load bpf program failed: ERROR: strerror_r(524)=22
libbpf: -- BEGIN DUMP LOG ---
libbpf: 
processed 80 insns (limit 1000000) max_states_per_insn 0 total_states 6 peak_states 6 mark_read 5

libbpf: -- END LOG --
libbpf: failed to load program 'kretprobe/tcp_v4_connect'
libbpf: failed to load object 'tcpconnect_bpf'
libbpf: failed to load BPF skeleton 'tcpconnect_bpf': -4007
failed to load BPF object: -4007
```

We are having some issues with perf on that system, so when we solve that and manage to debug the issue further I'll post a new comment with the updates.

**Comment by anakryiko** [comment]
> > We are currently testing that on the AP120C device. I will get back to you with the results tomorrow.
> > The patches to libbpf seem to have solved the previous verifier issue with 32/64 bit pointers.
> 
> Now we get the following results with ENOTSUPP again:
> 
> ```
> libbpf: Error in bpf_create_map_xattr(sockets):Invalid argument(-22). Retrying without BTF.
> libbpf: load bpf program failed: ERROR: strerror_r(524)=22

I think it's due to __atomic_add_fetch(). See my reply on mailing list (https://lore.kernel.org/bpf/CAEf4Bza2VA=eAOtmLaL23Fz07giN6AG3f5okwOdAcmrHda6AhQ@mail.gmail.com/), this is not supported on ARM32, unfortunately.

**Comment by dmarcinkovic** [comment]
> > > We are currently testing that on the AP120C device. I will get back to you with the results tomorrow.
> > > The patches to libbpf seem to have solved the previous verifier issue with 32/64 bit pointers.
> > 
> > 
> > Now we get the following results with ENOTSUPP again:
> > ```
> > libbpf: Error in bpf_create_map_xattr(sockets):Invalid argument(-22). Retrying without BTF.
> > libbpf: load bpf program failed: ERROR: strerror_r(524)=22
> > ```
> 
> I think it's due to __atomic_add_fetch(). See my reply on mailing list (https://lore.kernel.org/bpf/CAEf4Bza2VA=eAOtmLaL23Fz07giN6AG3f5okwOdAcmrHda6AhQ@mail.gmail.com/), this is not supported on ARM32, unfortunately.

I have tested the example without the BPF JIT and the error persists. Disabling the BPF JIT was the bug fix for the example on the mailing list, but it does not solve this issue.

**Comment by jvijtiuk** [comment]
> libbpf: Error in bpf_create_map_xattr(sockets):Invalid argument(-22). Retrying without BTF.
libbpf: load bpf program failed: ERROR: strerror_r(524)=22
libbpf: -- BEGIN DUMP LOG ---
libbpf: 
processed 80 insns (limit 1000000) max_states_per_insn 0 total_states 6 peak_states 6 mark_read 5
libbpf: -- END LOG --
libbpf: failed to load program 'kretprobe/tcp_v4_connect'
libbpf: failed to load object 'tcpconnect_bpf'
libbpf: failed to load BPF skeleton 'tcpconnect_bpf': -4007
failed to load BPF object: -4007

The remaining issues we had with tcpconnect weren't actually caused by tcpconnect. 
The kprobe crash on ARM32 is a generic issue with kprobes in ARM32, and it is fixed by this patch https://www.spinics.net/lists/kernel/msg3684757.html.

After applying that, we also applied this series of patches https://lore.kernel.org/bpf/20201007202946.3684483-1-andrii@kernel.org.

> I think it's due to __atomic_add_fetch(). See my reply on mailing list (https://lore.kernel.org/bpf/CAEf4Bza2VA=eAOtmLaL23Fz07giN6AG3f5okwOdAcmrHda6AhQ@mail.gmail.com/), this is not supported on ARM32, unfortunately.

Lastly, we did have to run the program without JIT, as suggested above. 

We also had some similar issues with runqslower on ARM32, which were fixed after the above steps as well.

tcpconnect now seems to be working on ARM32 so I'm closing the issue :)
